refactor: run integration tests outside LMS container

magajh opened the pull request:
**Overview**

This PR introduces significant modifications to the GitHub Action for testing Open edX plugins in the Tutor environment. These changes are necessary for improving the effectiveness and reliability of our integration tests

**Motivation for Changes**

The initial implementation ran integration tests inside the LMS container of Tutor. This approach created several challenges:
- **Networking Issues**: Running tests inside the same container restricted the ability to make requests to external hosts, leading to unreliable test outcomes.
- **Realistic Testing**: To ensure that our integration tests reflect real-world interactions between our plugins and the Open edX platform, it is crucial to run them outside the Tutor container, maintaining a separation between the tests and the platform environment.

To address these issues, this PR modifies the workflow to run integration tests outside the Tutor container, improving the testing environment and results.

**Workflow Steps and Their Necessity**

1. **Set Tutor Environment Variables**:
   - Sets necessary environment variables such as `LMS_HOST`, `CMS_HOST`, `TUTOR_ROOT`, and `TUTOR_PLUGINS_ROOT` for proper Tutor configuration.

2. **Install and Prepare Tutor**:
   - Installs the specified version of Tutor or clones the `nightly` branch if `tutor_version` is set to `"nightly"`.
   - Configures Tutor with the necessary host settings and launches it in non-interactive mode.

3. **Configure Caddyfile and Open edX Settings**:
   - Moved `patches.yml` into the action directory and copies it into the `plugins` directory during the workflow.
   - Enables the `patches` plugin to customize the Caddyfile and Open edX production settings. This configuration allows for the necessary routing for testing with multiple sites and security settings for our tests.

4. **Add Mount for Plugin**:
   - Mounts the plugin directory into the LMS and CMS containers, ensuring that the plugin code is accessible within the Tutor environment.

5. **Restart Tutor Services**:
   - Restarts Tutor services to apply the new mounts and configurations.

6. **Install Plugin and Extra Requirements**:
   - Installs the plugin in editable mode within the LMS container.
   - Installs any additional pip requirements specified in `openedx_extra_pip_requirements` in both LMS and CMS containers.

7. **Run Migrations**:
   - Executes database migrations for both the LMS and CMS to ensure the database schema is up-to-date.

8. **Import Demo Course**:
   - Imports the Open edX demo course for testing purposes.

9. **Set `DEMO_COURSE_ID` Environment Variable**:
  - Sets the DEMO_COURSE_ID environment variable based on the Tutor version.
  - Purpose: Allows tests to refer to the demo course ID when running integration tests.

10. **Test Open edX Imports in Plugin** *(Optional)*:
  - Provides a mechanism to verify that all imports from the Open edX platform in the plugin are working correctly.
  - Executes the test function specified by openedx_imports_test_file_path and openedx_imports_test_function_name, ensuring that the plugin's imports from Open edX do not have any issues.

11. **Load Initial Data for the Tests** *(Optional)*:
    - Loads initial data from a fixtures file into the LMS if `fixtures_file` is provided.

12. **Curl Heartbeat**:
    - Verifies that the LMS is running and responsive by checking the heartbeat endpoint.

13. **Run Integration Tests**:
    - Creates a virtual environment and runs the specified shell script to execute the integration tests outside the Tutor container, improving test reliability.

**Personalization Options**

The workflow includes several inputs that can be customized:

- `app_name`: The name of the application to test (e.g., `eox-core`).
- `tutor_version`: The version of Tutor to use, allowing for testing against specific versions or the `nightly` build.
- `shell_file_to_run`: The path to the shell script that contains the integration tests.
- `openedx_extra_pip_requirements`: Optional additional dependencies needed for testing.
- `fixtures_file`: Optional path to a JSON file containing initial data to load for the tests.
- `openedx_imports_test_file_path`: Optional path to the Python file containing the import test function.
- `openedx_imports_test_function_name`: Optional name of the function to execute in the import test file.

**Testing the Workflow**

To test this workflow:

1. **Create a Branch**:
   - Create a branch from any of the EOX plugins that includes integration tests. For example, you can use the branch [`bav/add-integration-tests` in `eox-core`](https://github.com/eduNEXT/eox-core/pull/275), which includes several integration tests.

2. **Add Integration Test Script**:
   - Include a shell script (e.g., `integration.sh`) in your plugin that installs dependencies and runs the tests.

     ```bash
     #!/bin/bash

     echo "Installing test requirements"
     make install-dev-dependencies

     echo "Running integration tests"
     pytest eox_core/api/v1/tests/integration/test_views.py
     ```

3. **Configure the Workflow**:
   - Update your workflow file to use the updated action and specify the required inputs.

     ```yaml
     name: Tutor Integration Tests
     on: [pull_request]

     jobs:
       integration-test:
         name: Tutor Integration Tests
         runs-on: ubuntu-latest
         strategy:
           matrix:
             tutor_version: ['<18.0.0', '==17.0.3', 'nightly']
         steps:
           - name: Run Integration Tests
             uses: eduNEXT/integration-test-in-tutor@v1
             with:
               tutor_version: ${{ matrix.tutor_version }}
               app_name: 'eox-core'
               shell_file_to_run: 'scripts/integration.sh'
               openedx_extra_pip_requirements: 'eox-tenant'
               fixtures_file: 'fixtures/initial_data.json'
               openedx_imports_test_file_path: 'eox_core/edxapp_wrapper/tests/integration/test_backends.py'
               openedx_imports_test_function_name: 'test_current_settings_code_imports'
     ```

4. **Verify the Tests Run Successfully**:
   - Ensure that the integration tests execute successfully in the workflow runs, validating that the action works as intended. You can refer to this test branch for an example: https://github.com/eduNEXT/eox-core/pull/288

**Additional Notes**

- **Supporting Tutor Nightly**:
  - The action now supports running tests against the `nightly` version of Tutor, enabling early detection of compatibility issues with upcoming changes.

- **Improved Documentation**:
  - The README has been updated to reflect the latest changes, including detailed input descriptions, usage examples, and special notes on using the `nightly` version.

magajh commented:
Note to reviewers: I'm working on updating the README

BryanttV commented on action.yml:
```diff
@@ -1,36 +1,119 @@
+    description: "Optional extra pip requirements to install in Open edX. E.g: 'package1==1.0 package2>=2.0'"
+    required: false
+    default: ""
+  python_version:
```
In which cases would this input (`python_version`) be useful?

magajh replied:
I couldn't find a case where it would be useful. I removed the step because I think it's better practice to use the default Python version from the runner. There's no really a need to install a specific Python version

Thanks!

BryanttV reviewed (commented):
@magajh, This looks great! thank you!

Just to note, when adding these changes we must modify all eox's that use this action, as the commands in your scripts will not work.

mariajgrimaldi commented on action.yml:
```diff
@@ -1,36 +1,119 @@
+      run: |
+        echo "Creating isolated venv to run the tests"
+        python -m venv .venv
+        if [ ! -d ".venv" ]; then
+          echo "Virtual environment creation failed"
+          exit 1
+        fi
```
Has the venv creation failed before? I can't think of why it could fail, but there may be edge cases I haven't considered. 

Also, should we create the environment before installing any dependencies like tutor?

magajh replied:
You're right. I think I initially added that check to ensure there were no issues with the creation, and then I forgot to remove it, but the step really never failed

Thanks!

mariajgrimaldi commented on action.yml:
```diff
@@ -1,36 +1,119 @@
+    - name: Configure Caddyfile and Open edX settings
+      run: |
+        mkdir -p plugins
+        cat << 'EOF' > plugins/patches.yml
+
+        name: patches
+        patches:
+          caddyfile: |
+            {$default_site_port} {
+              import proxy "lms:8000"
+            }
+          openedx-cms-production-settings: |
+            ALLOWED_HOSTS = ["*"]
+            ALLOWED_AUTH_APPLICATIONS = ['cms-sso', 'cms-sso-dev']
+          openedx-lms-production-settings: |
+            ALLOWED_HOSTS = ["*"]
+            ALLOWED_AUTH_APPLICATIONS = ['cms-sso', 'cms-sso-dev']
+        EOF
+        tutor plugins enable patches
```
Should we move this file to a Caddyfile outside the action definition, or would that be overkill? Composite actions can access their files, so maybe we can leverage that, although it might be more complex for little gain.

magajh replied:
Created a patch file that I'm importing in the step

magajh commented on action.yml:
```diff
@@ -1,36 +1,119 @@
+        echo "Copying fixtures file to the LMS container"
+        tutor local exec lms mkdir -p /openedx/fixtures
+        tutor local dc cp ${{ inputs.fixtures_file }} lms:/openedx/fixtures/fixture.json
+        tutor local exec lms python manage.py lms loaddata /openedx/fixtures/fixture.json
```
Thanks for noticing. I think using tutor mounts does make the execution look cleaner and avoids the overhead of copying files, which should, in theory, improve performance. But, I still need to stop and restart the services to make the mounts work, and this adds approximately an extra minute to the execution time
https://github.com/eduNEXT/integration-test-in-tutor/pull/4/files#diff-1243c5424efaaa19bd8e813c5e6f6da46316e63761421b3e5f5c8ced9a36e6b6R74
I could try just restarting the LMS and CMS containers

magajh replied:
Update: now I'm just restarting the LMS and CMS containers and the execution time is back to how it was before

magajh commented:
### Question for Reviewers

When support for the `nightly` version was added, it included building the Open edX image locally ([see line 45 in `action.yml`](https://github.com/eduNEXT/integration-test-in-tutor/blob/main/action.yml#L45)). However, this approach has led to significantly longer execution times and is causing resource exhaustion errors during the workflow runs, for example: https://github.com/eduNEXT/eox-core/actions/runs/11086481540/job/30804065328?pr=283

I'm thinking that building the image locally each time the tests are executed might be unnecessary. We can use the pre-built Tutor images that are published daily to test against the updates up to the time when the image was built, without the overhead of building images locally, which reduces execution time and prevents resource-related issues.

What are your thoughts on this approach?

magajh commented:
@BryanttV @mariajgrimaldi thank you for your comments, this is ready for review again

BryanttV commented on README.md:
```diff
@@ -14,30 +14,183 @@ jobs:
+- A comparison operator with a version (e.g., `"<18.0.0"`).
+- The string `"nightly"` to use the latest development version.
+
+*Examples*:
+
+- `"==17.0.3"`
+- `"<18.0.0"`
+- `"nightly"`
```
The same examples are shown above, could we use others, or leave it in one place?

magajh replied:
Done!

BryanttV commented on action.yml:
```diff
@@ -1,60 +1,188 @@
+    - name: Adjust permissions to execute Tutor commands
+      run: |
+        chmod 777 . -R
+      shell: bash
```
Could we leave a default shell as it's done in [this PR](https://github.com/eduNEXT/picasso/pull/24)?

magajh replied:
Done!

magajh replied:
It's not possible to use default shell in composite Actions :(
https://github.com/orgs/community/discussions/46670
https://github.com/orgs/community/discussions/18597

mariajgrimaldi commented on action.yml:
```diff
@@ -1,60 +1,188 @@
+
+    - name: Set Tutor environment variables
+      run: |
+        echo "LMS_HOST=local.edly.io" >> "$GITHUB_ENV"
+        echo "CMS_HOST=studio.local.edly.io" >> "$GITHUB_ENV"
+        echo "TUTOR_ROOT=$(pwd)" >> "$GITHUB_ENV"
+        echo "TUTOR_PLUGINS_ROOT=$(pwd)/plugins/" >> "$GITHUB_ENV"
```
Does [this](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/workflow-commands-for-github-actions#example-of-a-multiline-string) work?

```
      {
        LMS_HOST=local.edly.io
        CMS_HOST=studio.local.edly.io
        TUTOR_ROOT=$(pwd)
        TUTOR_PLUGINS_ROOT=$(pwd)/plugins/
        echo EOF
      } >> "$GITHUB_ENV"
```

magajh replied:
Did not worked exactly like this, but It worked using [Heredoc](https://github.com/eduNEXT/integration-test-in-tutor/pull/4/commits/839bffaf3eedf2bf3364188f848e7a0ab26e9e53). Thanks!

magajh replied:
That exact code didn't work, but using [heredoc](https://github.com/eduNEXT/integration-test-in-tutor/pull/4/commits/839bffaf3eedf2bf3364188f848e7a0ab26e9e53) did, which still helped me avoid repetition and made the code more readable. Thanks!

BryanttV commented on README.md:
```diff
@@ -14,30 +14,183 @@ jobs:
     strategy:
       matrix:
-        tutor_version: ["<17.0.0", "==17.0.3", "<18.0.0"]
+        tutor_version: ["<17.0.0", "==17.0.3", "<18.0.0", "nightly"]
```
I think this action doesn't work for versions `<17.0.0`. If so, we should mention it and remove these examples. Maybe, also mention that support is only guaranteed for the current version of Open edX, an earlier version, and nightly.

magajh replied:
yes, it's best to clarify. Thanks!

magajh replied:
Done

BryanttV commented on action.yml:
```diff
@@ -1,60 +1,188 @@
+        echo "TUTOR_PLUGINS_ROOT=$(pwd)/plugins/" >> "$GITHUB_ENV"
+      shell: bash
+
+    - name: Create virtualenvs
+      run: |
+        echo "Creating isolated venv for Tutor"
+        python -m venv .tutor_venv
+        echo "Creating isolated venv to run the integration tests"
+        python -m venv .tests_venv
```
Why do we need to use two environments?

magajh replied:
tutor_venv: This env is specifically for running Tutor to avoid permission problems. Without this, installing Tutor (especially the nightly version as an editable package) would try to install in the system-wide Python path, which causes problems with the permissions in the runner

tests_venv: This is used for executing the integration tests in a fully isolated environment. We do this because by separating the test environment, we prevent any potential conflicts with the Tutor environment and ensure that the tests run with only  the necessary dependencies

mariajgrimaldi replied:
So, using tutor_env fixes the issue, which is also solved by adjusting permissions [here](https://github.com/eduNEXT/integration-test-in-tutor/pull/4/files#diff-1243c5424efaaa19bd8e813c5e6f6da46316e63761421b3e5f5c8ced9a36e6b6R36). Or are those unrelated?

magajh replied:
@mariajgrimaldi  They are somewhat related but not interchangeable. Based on my testing while refactoring, here’s what I found:

The reason for using a venv for Tutor is to avoid permission issues when installing it as an editable package. Without the venv, `pip install -e` tries to install Tutor in the system-wide Python path, which requires root access. Changing permissions with `chmod` doesn’t affect system-wide directories (unless we do something like `sudo chmod 777 /usr/local/ -R`, which is not recommended), so that's why the venv is necessary in this case.

On the other hand, `chmod 777` is needed when installing the plugin as an editable package inside the Tutor containers. When running `pip install -e`, Python tries to create files like `*.egg-info` but doesn't have sufficient permissions to write to those directories inside the container

I can try narrowing the `chmod` to specific directories instead of applying it globally, which might help reduce the scope of the permission changes.

Here’s a link to the output when I only used `chmod` without the isolated env for Tutor installation:  
https://github.com/eduNEXT/eox-core/actions/runs/11133124056/job/30938507427?pr=283

And here’s the output when using the venv but without `chmod`:  
https://github.com/eduNEXT/eox-core/actions/runs/11133257345/job/30938936846?pr=283

I tried to get rid of one or the other (especially `chmod`, which I'd prefer to remove), but none of my attempts worked. I’ll open an issue so we can keep investigating and see if there’s a better solution

mariajgrimaldi replied:
Thank you so much for the thorough investigation, it's really helpful for future reference. For me, it's enough to leave this well documented in an issue, as you suggested.

mariajgrimaldi commented on action.yml:
```diff
@@ -1,60 +1,188 @@
+        tutor config save --set LMS_HOST=$LMS_HOST --set CMS_HOST=$CMS_HOST
+        tutor local launch -I
+      shell: bash
+      env:
+        INPUT_TUTOR_VERSION: ${{ inputs.tutor_version }}
```
I think moving this up after `name` would make the step easier to read.

magajh replied:
Done!

BryanttV commented on action.yml:
```diff
@@ -1,60 +1,188 @@
+        tutor local do importdemocourse
+      shell: bash
+
+    - name: Test Open edX imports in plugin
+      if: ${{ inputs.openedx_imports_test_file_path && inputs.openedx_imports_test_function_name }}
+      run: |
+        source .tutor_venv/bin/activate
+
+        cp $GITHUB_ACTION_PATH/run_openedx_imports_test.py $GITHUB_WORKSPACE/${{ inputs.app_name }}/
+        echo "Running test for Open edX imports in plugin"
+
+        tutor local exec --env PYTHONPATH=/openedx/${{ inputs.app_name }} \
+                        --env DJANGO_SETTINGS_MODULE=lms.envs.production \
+                        lms python /openedx/${{ inputs.app_name }}/run_openedx_imports_test.py \
+                        --test-file-path /openedx/${{ inputs.app_name }}/${{ inputs.openedx_imports_test_file_path }} \
+                        --test-function-name '${{ inputs.openedx_imports_test_function_name }}'
```
I remember we used a step similar to this to test the imports, isn't it possible to do it this way now?
```yaml
steps:
  - name: Run backend tests
      if: ${{ inputs.backends_test }}
      run: |
        tutor local exec lms pytest -s --ds=lms.envs.tutor.test /openedx/eox-plugin/path/to/tests/test_backends.py 
      shell: bash
```

magajh replied:
We would need to first create a venv and install the test dependencies
Please check this comment https://github.com/eduNEXT/integration-test-in-tutor/pull/4#discussion_r1783333738

mariajgrimaldi commented on action.yml:
```diff
@@ -1,60 +1,188 @@
+
+        tutor local exec --env PYTHONPATH=/openedx/${{ inputs.app_name }} \
+                        --env DJANGO_SETTINGS_MODULE=lms.envs.production \
+                        lms python /openedx/${{ inputs.app_name }}/run_openedx_imports_test.py \
+                        --test-file-path /openedx/${{ inputs.app_name }}/${{ inputs.openedx_imports_test_file_path }} \
+                        --test-function-name '${{ inputs.openedx_imports_test_function_name }}'
```
Should we use environment variables for these long paths?

```
env:
     OPENEDX_IMPORTS_TEST_PATH: /openedx/${{ inputs.app_name }}/run_openedx_imports_test.py
     TEST_FILE_PATH: /openedx/${{ inputs.app_name }}/${{ inputs.openedx_imports_test_file_path }}
```

magajh replied:
Much better. Thanks

mariajgrimaldi commented on action.yml:
```diff
@@ -1,60 +1,188 @@
+      with:
+        path: ${{ inputs.app_name }}
+
+    - name: Adjust permissions to execute Tutor commands
+      run: |
+        chmod 777 . -R
```
This question is totally out of scope, considering it was a design decision made and approved in a previous PR. Please feel free to reference any previous discussions in case I've forgotten or missed them.

We had [this workflow](https://github.com/eduNEXT/tutor-local-nightly-athena/blob/main/.github/workflows/build.yml) to build nightly images that didn't need permission adjustments. What's the difference between this GH action and the workflow?

mariajgrimaldi replied:
Please, see: https://github.com/eduNEXT/integration-test-in-tutor/pull/4#discussion_r1783558868

mariajgrimaldi commented:
@magajh: thanks for the thorough explanation! Pulling the image instead of building it makes total sense to me.

BryanttV commented on action.yml:
```diff
@@ -1,60 +1,188 @@
+        chmod +x ./${{ inputs.shell_file_to_run }}
+        ./${{ inputs.shell_file_to_run }}
+      shell: bash
+      env:
+        DEMO_COURSE_ID: ${{ env.DEMO_COURSE_ID }}
+        LMS_HOST: ${{ env.LMS_HOST }}
+        CMS_HOST: ${{ env.CMS_HOST }}
```
Why do we need to add these env variables again?

magajh replied:
I was testing something and forgot to remove them. Thanks for noticing!

magajh commented on README.md:
```diff
@@ -14,30 +14,183 @@ jobs:
+The name of the function in the specified file that executes the import tests for Open edX.  
+*Example*: `"test_openedx_imports"`
+
+## Description
```
Thanks for the suggestion! I think starting with the Usage section is better in general because it lets users quickly understand how to implement the action without diving into detailed explanations right away. They can always refer to the Description section later for more information. What I like about this structure is that we basically prioritize actionable steps while still providing detailed information for those who want it.

Let me know what you think

mariajgrimaldi replied:
That makes sense. We could change the title from description to overview or something similar so the actual description of the ga is in the first couple of lines of the document.

magajh replied:
Agreed, I've changed the name to _Overview_. Thanks!

magajh commented on action.yml:
```diff
@@ -1,60 +1,188 @@
+
+        tutor local exec lms python manage.py lms migrate
+        tutor local exec cms python manage.py cms migrate
+        tutor local restart
```
Good catch, thanks!

BryanttV replied:
> Maybe we're missing a restart for an additional service (?)

Maybe the workers? :thinking:

magajh replied:
tried to restart the workers as well and the same problem appears https://github.com/eduNEXT/eox-core/actions/runs/11152363622/job/30997831749?pr=283
It really doesn't affect the execution time so I think we can just do restart

BryanttV replied:
I find it very strange that it doesn't work, but I agree to restart all of them.

magajh commented on README.md:
```diff
@@ -14,30 +14,183 @@ jobs:
-(Optional) The path of the shell file to run the integration tests.
+**Required**  
+The path to the shell script that runs your integration tests. This path is relative to your plugin directory.  
+*Example*: `"tests/integration.sh"`
```
Done. Thanks!

magajh commented on README.md:
```diff
@@ -14,30 +14,183 @@ jobs:
+The path to the fixtures file of your plugin to load initial data for the tests. This path is relative to your plugin directory.  
+*Example*: `"fixtures/test_data.json"`
+
+### `openedx_imports_test_file_path`
+
+**Optional**  
+The path to the Python file in your plugin that contains the test function for validating Open edX imports. This path is relative to your plugin directory.  
+*Example*: `"tests/import_tests.py"`
+
+### `openedx_imports_test_function_name`
+
+**Optional**  
+The name of the function in the specified file that executes the import tests for Open edX.  
+*Example*: `"test_openedx_imports"`
```
@mariajgrimaldi, @BryanttV: 

We previously used pytest to run these tests when the action was executed inside the LMS container.
Initially, I considered using pytest within the LMS container for this test as well, but this approach would require setting up a new virtual environment inside the container and installing test dependencies like pytest and pytest-django. After checking the import test and realizing is a simple function that makes the imports, I thought it might unnecessary setting up pytest when we could just directly invoke the function within the container

I'm interested in hearing your thoughts on this approach

mariajgrimaldi replied:
Thanks for the additional context, now I understand the implementation better. My only concern with this approach is that maintaining it would mean maintaining another way of running tests besides the Pytest approach. Does installing Pytest add so much overhead?

BryanttV replied:
Thanks @magajh. I agree with Majo's question. Also, do you happen to have any log for the jobs that shows what it looks like when some backend fails?

magajh replied:
@mariajgrimaldi @BryanttV: I agree that running tests in two different ways might not be ideal, and we should stick to only executing the tests with pytest if possible

Here’s an update on the progress I’ve made while setting up pytest:

After testing further, I found that installing `pytest` and `pytest-django` dependencies directly in the container works perfectly (as I did in the [latest commit](https://github.com/eduNEXT/integration-test-in-tutor/pull/4/commits/935eafedd0f2529a80e61e8087606af6793e5885)). However, I don’t think this is the best approach since we are permanently altering the container and affecting the environment where the platform runs—something we want to avoid.

I’ve been trying to create a separate virtual environment to install these dependencies, but that’s where I’ve encountered issues. The virtual environment created for testing does not inherit the globally installed packages in the LMS container, such as Django. I believe this is because in the Docker container, the packages might be installed in paths that aren’t accessible or visible to the new virtual environment. 

I tried using the `--system-site-packages` option when creating the venv to make it recognize global packages, but it didn’t work. So, I ended up installing all the dev dependencies using the target in the Makefile (`install-dev-dependencies`), which worked fine. The problem I still can’t solve is that running pytest in that venv doesn’t work. The key issue seems to be that the virtual environment inside the LMS container is not properly set up to recognize the LMS modules and paths needed for the tests.

I’ve been trying to adjust the `PYTHONPATH` to include the necessary directories when running the tests, but so far, it hasn’t worked.

Here’s an example of the error I’m getting: https://github.com/eduNEXT/eox-core/actions/runs/11150314446/job/30991166191#step:2:7014

I’ll try a few more things to get pytest running in an isolated venv and will update you here.

In case you want to check the branch with the workflow where I’m trying to make this work in the venv, it’s `mjh/test-changes-run-tests-outside-container`

BryanttV replied:
Thank you for all the context, @magajh! Have you tried using [`tutor local run`](https://docs.tutor.edly.io/reference/cli/local.html#tutor-local-run)? This command creates a new container, and we would not be affecting the one already running, but I'm not sure it works.

mariajgrimaldi replied:
> The problem I still can’t solve is that running pytest in that venv doesn’t work. The key issue seems to be that the virtual environment inside the LMS container is not properly set up to recognize the LMS modules and paths needed for the tests.

Did you try running `pip install -e .` inside the venv?

Also, pytest or pytest-django don't share dependencies with what's already installed in the venv for a production environment (base.txt). From what I gathered from the repositories, pytest-django only [requires pytest](https://github.com/pytest-dev/pytest-django/blob/1ffc3239387bcfadbd8aecfdfacf3d6b52b947f8/pyproject.toml#L40-L42), and [pytest requires a list of libraries](https://github.com/pytest-dev/pytest/blob/7add2af7087dc32af19a38f5950a5b64a19d1de5/pyproject.toml#L49-L56) without any specific versions -- at least that conflict with our production environment, some are installed for testing or development, but not in base.txt.

magajh replied:
@BryanttV since `tutor local run` creates a new container from the image, we would need to install the plugin, the extra pip requirements, run migrations, execute the test for the backends - all in the same run command -, and then do it again with `tutor local exec` to run the integration tests

mariajgrimaldi replied:
I agree with leaving it as is with pytest. Thank you for addressing all of our concerns!

magajh replied:
Updated the readme and workflow to remove any trace of the `openedx_imports_test_function_name` input

magajh commented:
@mariajgrimaldi @BryanttV: thank you for your valuable feedback! I've addressed all the comments and this is ready for review again. I think the only pending discussion that we have is https://github.com/eduNEXT/integration-test-in-tutor/pull/4#discussion_r1783333738

BryanttV commented on run_openedx_imports_test.py:
```diff
@@ -0,0 +1,49 @@
+
+import argparse
+import importlib.util
+import os
```
Unused import
```suggestion
```

magajh commented on README.md:
```diff
@@ -14,30 +14,181 @@ jobs:
+**Optional**  
+The path to the shell script that runs your integration tests. This path is relative to your plugin directory. 
+*Default*: `"scripts/execute_integration_tests.sh"`
+*Example*: `"tests/integration.sh"`
```
Done

BryanttV reviewed (approved):
This looks great! Thank you very much for addressing our comments!
